### PR TITLE
Add option to limit the number of lines used for guessing in io.ascii

### PR DESCRIPTION
**Pull request by astrofrog**
I ran into the issue again today that when reading a large table with io.ascii, the guessing is very slow because each format uses the whole table to guess the format. However, in many cases, it is usually likely to be good enough to use the first say 1000 lines to guess the format. The approach I've tried here is that in the guessing, we first try the first N lines of the file, and if it works we try again with the full table to make sure, but this allows many non-matches to exit early.

Currently, this is opt-in, but with a warning for users not setting ``guess_limit_lines`` if the table is reasonably large, though arguably we may want to set a sensible default, say 1000 or 10000 lines by default, so that this benefits most users, instead of having the warning. Example usage with a table with 750k lines:

```python
In [1]: from astropy.io.ascii import read
   ...: 

In [2]: %time read('test.csv')
/home/tom/Code/astropy/astropy/io/ascii/ui.py:668: UserWarning: Input file contains many lines,
so guessing the format may be slow. Consider setting ``guess_limit_lines``to limit the number of
lines to use for guessing the format
  warnings.warn(
CPU times: user 11.7 s, sys: 886 ms, total: 12.6 s
Wall time: 12.6 s
Out[2]: 
<Table length=786432>
         x                    y                    z                  l                  b                  d         
      float64              float64              float64            float64            float64            float64      
-------------------- -------------------- ------------------- ------------------ ------------------ ------------------
  0.4363633036730489   0.4363633036730488   193.4849355151462               45.0  89.81725848475484  193.4859196354243
-0.43813116612984976   0.4381311661298498   194.2688115894507              135.0  89.81725848475484 194.26979969674898
...

In [3]: %time read('test.csv', guess_limit_lines=1000)
CPU times: user 1.08 s, sys: 32.1 ms, total: 1.11 s
Wall time: 1.11 s
Out[3]: 
<Table length=786432>
         x                    y                    z                  l                  b                  d         
      float64              float64              float64            float64            float64            float64      
-------------------- -------------------- ------------------- ------------------ ------------------ ------------------
  0.4363633036730489   0.4363633036730488   193.4849355151462               45.0  89.81725848475484  193.4859196354243
-0.43813116612984976   0.4381311661298498   194.2688115894507              135.0  89.81725848475484 194.26979969674898
...
```

@taldcroft if you are on board with this kind of approach, I can add tests and docs, but didn't want to waste too much time if you think it is not going to be robust. Let me know what you think about whether we should have this be opt-in with a warning as done here, or opt-out?

Fixes https://github.com/astropy/astropy/issues/7977 
Related: https://github.com/astropy/astropy/issues/7976

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by taldcroft** [comment]
@astrofrog - this seems like a generally good idea, so :+1: on moving forward. I have a couple of ideas to discuss.

First, the `read()` API and guessing are already very complex. I would propose that we can make line-limiting be the default. Since changing the line limit would be a rare situation, we could have the value set as a config parameter.

Second, I'm wondering if the more useful limit is the number of characters since that is more closely proportional to read time. This is a less intuitive but probably does provide a more even limit on the time spent reading. Algorithmically you would chop to a certain number of characters and then back up to the first `\n` (I think?). Doing this also means you don't need to count lines and it is easy to skip the preview guess read if the table is under the character limit. I'm not adamant about this idea but it's worth thinking about.

One other thing is fixing the `_guess` docstring and/or adding type annotations. The docs say that `table` can be a str or list or file object (which would break everything), but I think it really has to be a str.

**Comment by taldcroft** [comment]
One more thing to consider is some hysteresis. Using the current line limit implementation, if a table is 1001 lines long with a limit of 1000 then you are better off just skipping the preview read. So there is another configuration variable like some factor (e.g. 2) where the preview step is skipped if actual lines is less than limit_lines * factor. This idea applies exactly the same if the limit is number of characters.

It would probably also be nice (if possible) to factor out all this into a single function like `read_preview(table, ...)` that returns nothing but can raise an exception if reading the table fails.

**Comment by astrofrog** [comment]
@taldcroft - thanks, that all makes sense! I hesitated between limiting by lines or characters, as my worry is that one could imagine a table with very long lines and we do want to make sure we allow enough lines for the guessing. Do you think that in addition to limiting by characters we should ensure a minimum number of actual lines? If so, should that be a separate config item?

**Comment by pllim** [comment]
If the API is user facing, lines would be more user-friendly. FWIW

File: astropy/io/ascii/ui.py [hunk 1]
```diff
@@ -641,6 +645,33 @@ def _guess(table, read_kwargs, format, fast_reader):
             reader = get_reader(**guess_kwargs)
 
             reader.guessing = True
+
+            # Start off by checking what line endings are being used in file
```
Comment (mhvk) [inline astropy/io/ascii/ui.py:649]: This should be moved outside the loop so that one doesn't find the short piece for every table type.

**Review by mhvk — COMMENTED** [review]
:+1: to the whole idea, and also :+1: to making this the default. I guess in principle it would be nicer if the individual readers just failed fast (i.e., #7976 for all readers), but that is a much larger effort, while this one provides a generic benefit. 

I guess my only worry with this is that there may be table types for which the header is very long -- 1000 lines for `.ecsv` would not seem impossible... But perhaps it is fine if for unusual tables users to intervene and just tell what the format is.

**Comment by mhvk** [comment]
p.s. I would say this does not fix #7976 - if the `.ecsv` reader fails without the `%ECSV` line, then that should be the first thing to check.

**Comment by astrofrog** [comment]
Perhaps if there are no matches we could call _guess again with no limits?

**Comment by pllim** [comment]
I modified the OP to not close the issue that Marten said this isn't fixing. FYI

**Comment by taldcroft** [comment]
> I guess my only worry with this is that there may be table types for which the header is very long -- 1000 lines for .ecsv would not seem impossible... But perhaps it is fine if for unusual tables users to intervene and just tell what the format is.

This is a good point and we need to be careful. What would happen here is that the `ecsv` reader would fail and then it would fall through to a lower-priority reader like `csv`, which would succeed since the data are valid CSV. From the user perspective there would be no error and the table would look approximately correct.

ECSV is an example of a reader that does fail fast if the header does not conform to the spec. So for that format (and others that have stringent header specifications which can short-circuit reading), we would not want to limit the lines. That, of course, means this simple enhancement is getting more complex... But maybe not too bad, perhaps some class attribute that indicates that it has short-circuit parse failure detection.

**Comment by mhvk** [comment]
I like the idea that the class itself can tell whether it has "fast fail"!

**Comment by taldcroft** [comment]
I took a quick look at short circuiting in general assuming the current `io.ascii` flow where it always reads the entire file into memory first (unless fast-ascii chunking is specified). Reading only a part of the file initially would be a bigger change (possible, but bigger).

The low-hanging fruit here is to check for each data row that the number of parsed data columns matches the number of header columns. The fast reader already does this and jumps out at the first failure (at least based on my read of the code, this needs checking). The Python basic readers do not do this, but at first glance it seems like it would be quite easy.

Other things can go wrong in a reader, like the data don't match the expected data type, but this is much less common.

**Comment by taldcroft** [comment]
I discovered another bottleneck in the current implementation which is that each Python reader independently calls a function to split the text into lines. This would require bigger structural changes to circumvent.

The upshot is that the approach of limiting lines in this PR is likely to be the most effective "easy" approach. I'm still not sure if limiting lines or characters is best, but maybe conceptually lines is more intuitive and we should just stick with that.

**Comment by astrofrog** [comment]
@taldcroft @mhvk - I've now implemented your comments/suggestions related to this PR, and it is now ready for a proper review

File: astropy/io/ascii/ui.py [hunk 2]
```diff
@@ -641,6 +665,13 @@ def _guess(table, read_kwargs, format, fast_reader):
             reader = get_reader(**guess_kwargs)
 
             reader.guessing = True
+
+            if table_guess_subset:
```
Comment (taldcroft) [inline astropy/io/ascii/ui.py:669]: I think this is the spot where I think we want something like
```
if table_guess_subset is not None and getattr(reader, "enable_guess_limit_lines"):
```
Then the `Reader` class for any class where applying a guess limit is safe should define a class attribute `enable_guess_limit_lines = True`. So this would include all the `Basic` readers but e.g. `ECSV`, `CDS`, `IPAC` would NOT have this (or set it to `False`).

I originally imagined something to opt out, but considering there could be user-defined classes out there where there is a non-trivial header and the potential for something silly like a user configuring to check only the first 100 lines and shooting themself in the foot.

One interesting case is the C-parser fast readers. From my testing and look at the code, I think they short-circuit quickly because there is a line-by-line check that the number of data columns matches the number of header names. So we might be better off leaving those out.

**Comment by taldcroft** [comment]
One credible case is where a user specifies `data_start=20000` to skip ahead. I think that if the user provides any of `header_start`, `data_start`, or `data_stop` to the `read()` call then line limiting should be disabled. Otherwise things start getting ambiguous.

File: astropy/io/ascii/ui.py (hunk 2 above)
Comment (astrofrog) [reply astropy/io/ascii/ui.py:669]: So is the idea that readers that can fail fast anyway should not be included here? Note that for ECSV we need e.g. https://github.com/astropy/astropy/pull/16850#event-13930141080 to truly fail fast and not start processing all the lines.

**Comment by astrofrog** [comment]
Ok sounds good, I'll try and work on this some more!

**Comment by pllim** [comment]
tl;dr -- If downstream package (e.g., specutils) registers their own I/O formats. Do they need to do anything special to use (or not use) this new feature?

**Comment by taldcroft** [comment]
> tl;dr -- If downstream package (e.g., specutils) registers their own I/O formats. Do they need to do anything special to use (or not use) this new feature?

They would need to add a class attribute to their format(s) to opt in to line limits in guessing. If they do nothing then nothing changes.

**Comment by mhvk** [comment]
An opt-in method would seem good. Inspired by #16850, that perhaps should be the presence of a `.validate` method or so, which can then get called if it exists and is `callable` (and may have other meanings if it is not callable). Maybe also be good to get @nstarman's opinion, as he has some type-recognition things in `cosmology`.

But also good not to let the perfect be the enemy of a great improvement!

**Comment by nstarman** [comment]
@mhvk. Sorry, can you clarify?

**Comment by mhvk** [comment]
@nstarman - I pinged you as in cosmology you have methods to recognize various formats quickly, which is what would seem useful here too.

**Comment by nstarman** [comment]
Oh! we have `Cosmology.from_format`, which is just Astropy's universal I/O layer, like `Table.read`. So I think it's the same thing as here!

File: astropy/io/ascii/ui.py (hunk 2 above)
Comment (taldcroft) [reply astropy/io/ascii/ui.py:669]: I am basically rescinding my above comment, but would like to leave it for the record.

File: docs/changes/io.ascii/16840.perf.rst [hunk 3]
```diff
@@ -0,0 +1,4 @@
+Guessing of table formats when reading is now based by default on the first
+10000 lines of a file (if it is longer than this). This behavior can be
+configured using the ``astropy.io.ascii.conf.guess_limit_lines`` configuration
+item.
```
Comment (taldcroft) [inline docs/changes/io.ascii/16840.perf.rst:4]:
```suggestion
The performance of guessing the table format when reading large files with 
``astropy.io.ascii`` has been improved. Now the process uses at most 
10000 lines of the file to check if it matches the format. This behavior can 
be configured using the ``astropy.io.ascii.conf.guess_limit_lines`` 
configuration item, including disabling the limit entirely.
```

File: docs/whatsnew/7.0.rst [hunk 4]
```diff
@@ -123,6 +124,24 @@ value in ``Unit.to`` and have those arrays not be converted to Numpy arrays::
...
 Note that it is not yet possible to use ``Quantity`` with dask arrays directly.
 
+.. _whatsnew_7_0_faster_io_ascii_guessing:
+
+Faster guessing of formats in ``astropy.io.ascii``
+==================================================
+
+When reading text-based tables with ``astropy.io.ascii`` and without specifying the
+table format, guessing the format should now be faster for large tables as it is now
+based on a subset of the table - by default the first 10000 lines of the file. If
```
Comment (taldcroft) [inline docs/whatsnew/7.0.rst:134]: I think some slight variation of my suggestion for the change log would work here to make a more direct statement about the performance improvement.

File: docs/whatsnew/7.0.rst [hunk 5]
```diff
@@ -123,6 +124,24 @@ value in ``Unit.to`` and have those arrays not be converted to Numpy arrays::
...
+needed, this limit can be modified with the following configuration item:
+
+.. code-block:: python
+
+    >>> from astropy.io.ascii import conf
+    >>> conf.guess_limit_lines = 100000
+
+and the limit can be disabled by setting this value to `None`. For large tables, it
+is however preferable to specify the format of the table explicitly when reading.
```
Comment (taldcroft) [inline docs/whatsnew/7.0.rst:143]:
```suggestion
and the limit can be disabled by setting this value to `None`. When reading large tables
you should specify the format of the table explicitly if possible.
```

**Review by taldcroft — CHANGES_REQUESTED** [review]
I made some text suggestions and it looks like this needs a rebase. Otherwise this looks basically good to me. I still need a chance to try out the code and think about corner cases, but let's plan on getting this in to 7.0. I'm bailing on #16860 for now.

File: docs/whatsnew/7.0.rst (hunk 4 above)
Comment (astrofrog) [reply docs/whatsnew/7.0.rst]: Done

**Comment by astrofrog** [comment]
@taldcroft - comments implemented and rebased

File: astropy/io/ascii/ui.py [hunk 6]
```diff
@@ -627,6 +627,30 @@ def _guess(table, read_kwargs, format, fast_reader):
         cparser.CParserError,
     )
 
+    # Check what line endings are being used in file
+    line_ending = "\r\n" if "\r" in table else "\n"
```
Comment (taldcroft) [inline astropy/io/ascii/ui.py:631]: In the more common case of `\n` line endings, this requires scanning the entire text. While that is relatively fast, is this handling of the line ending really necessary? 

If the later code just does `table.index("\n", pos + 1)`, then it seems like that will appropriately count lines (in either case) and then on Windows just end up with a dangling `\r`. I *think* that is OK and it seems to work, but I'm not sure if there is some subtlety I'm not thinking about.
```
In [4]: text = "a b\r\n1 2\r"

In [5]: ascii.read(text)
Out[5]: 
<Table length=1>
  a     b  
int64 int64
----- -----
    1     2
```

**Review by taldcroft — COMMENTED** [review]
I made a couple of comments and suggestions. Otherwise this looks good.

**Comment by neutrinoceros** [comment]
@astrofrog could you rebase again to clear CI, please ?

File: astropy/io/ascii/ui.py (hunk 6 above)
Comment (astrofrog) [reply astropy/io/ascii/ui.py]: Fixed

**Comment by astrofrog** [comment]
@taldcroft - I realised that ``table`` can also be a list of strings or a file object. I have fixed the easy case where it is a list of lines, but I was wondering about the file object case. Interestingly this doesn't seem to be covered by the test suite - is it that at this point in the code ``table`` won't ever be a file object? Or is this a deficiency in the test suite?

**Comment by taldcroft** [comment]
> @taldcroft - I realised that `table` can also be a list of strings or a file object. I have fixed the easy case where it is a list of lines, but I was wondering about the file object case. Interestingly this doesn't seem to be covered by the test suite - is it that at this point in the code `table` won't ever be a file object? Or is this a deficiency in the test suite?

Good catch. I had assumed that `get_readable_fileobj` would work with a text file object so that the downstream code would not see a fileobj:
```
            try:
                table = _expand_user_if_path(table)
                with get_readable_fileobj(table, encoding=encoding) as fileobj:
                    table = fileobj.read()
            except ValueError:  # unreadable or invalid binary file
                raise
            except Exception:
                pass
```
Alas `get_readable_fileobj` seems to only accept a bytes file object:
```
In [27]: fh = open("junk.dat", "r")

In [28]: with get_readable_fileobj(fh, encoding="utf8") as fileobj:
    ...:     fileobj.read()
    ...: 
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[28], line 2
      1 with get_readable_fileobj(fh, encoding="utf8") as fileobj:
----> 2     fileobj.read()

TypeError: read() should return bytes
```
That means the code is indeed sending a file object to `_guess` via the `except Exception` passthrough. Hmm. The logic here is not really straightforward.

Can you disable the guess lines limit for the case of a file object?

**Comment by astrofrog** [comment]
@taldcroft - I've skipped the limit limiting now for file objects. I ran into issues when trying to add a test for this because it seems that even something simple like:

```
def test_guess_with_fileobj():
    with open("data/ipac.dat", 'r') as f:
        ascii.read(f)
```

doesn't seem to work properly, unless I'm missing something?

**Comment by taldcroft** [comment]
The new code looks fine in principle, but I note that lots of tests are failing and it appears related.

> doesn't seem to work properly on main, unless I'm missing something?

Yck. But apparently nobody does this since that code is really old.

**Comment by dhomeier** [comment]
> > doesn't seem to work properly on main, unless I'm missing something?
> 
> Yck. But apparently nobody does this since that code is really old.

But even `simple.txt` with `format` specified does not work on the file object, that's not good!

**Comment by dhomeier** [comment]
> def test_guess_with_fileobj():
>     with open("data/ipac.dat", 'r') as f:
>         ascii.read(f)

Ah, you need to set 'rb', because `get_readable_fileobj` requires binary mode. Has some potential for confusion!

**Comment by astrofrog** [comment]
We should probably check that and raise a more explicit error though right?

Is there no way we could support files opened with r?

**Comment by dhomeier** [comment]
I see @taldcroft already noted that above – but is there any specific functionality from text mode that you'd need here?

The exception being reported as 
> TypeError: Input "table" must be a string (filename or data) or an iterable

indeed should be fixed.

File: astropy/io/ascii/ui.py [hunk 7]
```diff
@@ -627,6 +627,37 @@ def _guess(table, read_kwargs, format, fast_reader):
...
+    # there are just above the number of lines compared to the limit, up to a
+    # factor of 2, since it is fast to just go straight to the full table read.
+    table_guess_subset = None
+
+    if limit_lines:
+
+        if isinstance(table, list):
+            if len(table) > 2 * limit_lines:
+                table_guess_subset = table[:limit_lines]
+        else:
```
Comment (dhomeier) [inline astropy/io/ascii/ui.py:648]:
```suggestion
        elif isinstance(table, io.BufferedReader):
            line_len = len(table.readline())
            table.seek(0)
            table_guess_subset = table.readlines(line_len * limit_lines)
            table.seek(0)
        else:
```
Kind of a crude guess; not sure if this would have to be wrapped in `get_readable_fileobj` on top.

File: astropy/io/ascii/ui.py [hunk 8]
```diff
@@ -627,6 +627,37 @@ def _guess(table, read_kwargs, format, fast_reader):
         cparser.CParserError,
     )
 
+    # Determine whether we should limit the number of lines used in the guessing.
+    # Note that this does not currently work for file objects, so we set this to
+    # False if a file object was passed in.
+    from astropy.io.ascii import conf  # avoid circular imports
+
+    limit_lines = conf.guess_limit_lines and not hasattr(table, 'read')
```
Comment (dhomeier) [inline astropy/io/ascii/ui.py:635]:
```suggestion
    limit_lines = conf.guess_limit_lines * not hasattr(table, 'read')
```
or
```suggestion
    limit_lines = conf.guess_limit_lines if not hasattr(table, 'read') else False
```
This was setting `limit_lines = True`, which caused only the first line without newline to be read into `table_guess_subset`.

**Comment by astrofrog** [comment]
Need to fix typo in commit message...

**Comment by dhomeier** [comment]
... and a trailing space in the docs 😬

File: astropy/io/ascii/ui.py (hunk 7 above)
Comment (astrofrog) [reply astropy/io/ascii/ui.py:647]: I'd rather leave supporting file objects to another PR - in particular I need to understand if some file objects are streaming and don't support ``seek(0)`` for instance.

Comment (dhomeier) [reply astropy/io/ascii/ui.py:647]: Makes sense. `get_readable_fileobj` does check this and I believe, is making sure the returned file object provides `seek(0)`, but at that point may already have created more overhead than directly reading the file in.
https://github.com/astropy/astropy/blob/3099d86fbf0f7f88ed079fd50a563ad661fa2ca1/astropy/utils/data.py#L376-L386

File: astropy/io/ascii/tests/test_read.py [hunk 9]
```diff
@@ -2137,3 +2137,40 @@ def test_table_write_help_ascii_html():
     assert "Parameters" in doc
     assert "ASCII writer 'ascii.html' details" in doc
     assert "**htmldict** : Dictionary of parameters for HTML input/output." in doc
+
+
+def test_table_guess_limit_lines():
```
Comment (taldcroft) [inline astropy/io/ascii/tests/test_read.py:2142]: The test is not covering the case of an input table as a list of lines. There are 5 types that can be supplied to `read`, and we are ignoring file-like. I think this test function can be parametrized with a `table_type = ["filename", "list", "table_text", "path"]` and then an if-else or switch to get the appropriate `table` object from the `data/ipac.dat` file.
```
    table : str, file-like, list, `pathlib.Path` object
        Input table as a file name, file-like object, list of string[s],
        single newline-separated string or `pathlib.Path` object.
```
This also means doing the `data/sextractor2.dat` as a separate test.

**Review by taldcroft — CHANGES_REQUESTED** [review]
One last change request for testing, otherwise this looks good to me.

File: astropy/io/ascii/tests/test_read.py (hunk 9 above)
Comment (astrofrog) [reply astropy/io/ascii/tests/test_read.py]: Done!

**Comment by astrofrog** [comment]
@taldcroft - I've parameterized the test as you suggested and split out the sextractor test. Does this all seem ok now?

**Review by taldcroft — APPROVED** [review]
Good, woohoo!
